### PR TITLE
Divergence-free rewards distance calculation

### PR DESCRIPTION
**Pull request by anshradh**
Builds off #1 to implement divergence-free reward distance calculation.

### CODE REVIEW

File: src/epic/distances/divergence_free.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        rew_fn = utils.multidim_rew_fn(reward_function)
+        n_samples_can = n_samples_can or self.default_samples_can
+        assert isinstance(n_samples_can, int)
+
+        net = nn.Sequential(
+            *[
+                nn.Flatten(),
+                nn.Linear(self.state_dim, max(self.state_dim * 4, 128)),
+                nn.GELU(),
+                Residual(nn.Linear(max(self.state_dim * 4, 128), max(self.state_dim * 4, 128))),
```
Comment (ejnnr) [inline src/epic/distances/divergence_free.py:81]: I think people usually have residual connection over multiple layers (e.g. two linear layers plus the non-linearity in between)? If you just have a linear layer, I wouldn't expect there to be much of a benefit (since it's easy for linear layers to fit an identity), but maybe I'm wrong about that. Anyway, this is just a thought on things to try if the optimization isn't working, I'm aware this specific network is temporary and we'll likely have something more flexible later

File: src/epic/distances/divergence_free.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                nn.Linear(max(self.state_dim * 4, 128), 1),
+            ],
+        )
+
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        net.to(device)
+
+        max_epochs = 10000
+        optimizer = optim.AdamW(net.parameters(), lr=6e-4)
+        scheduler = optim.lr_scheduler.LambdaLR(
```
Comment (ejnnr) [inline src/epic/distances/divergence_free.py:92]: In my experience with similar optimization problems (though in a different project), a learning rate scheduler isn't usually necessary to get things to work at all (though it may help slightly). So for testing purposes, I'd probably try to make things work without the scheduler, just because that adds more hyperparameters to fiddle with. But again, those are just my thoughts on debugging, in principle having an LR scheduler seems good (and if it helps you make things work, then nvm my comment)

File: src/epic/distances/divergence_free.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        losses = []
+
+        with torch.inference_mode():
+            if state_sample_tensor.ndim == 1:
+                assert next_state_sample_tensor.ndim == 1
+                state_sample_tensor.unsqueeze_(-1)
+                next_state_sample_tensor.unsqueeze_(-1)
+
+            potential = (self.discount_factor * net(next_state_sample_tensor) - net(state_sample_tensor)).squeeze(-1)
```
Comment (ejnnr) [inline src/epic/distances/divergence_free.py:111]: Note on terminology: "potential" means the function on states itself, i.e. the network in our case. This term with a difference of two potentials would be "potential shaping", "potential shaping term", "pure potential shaping", or maybe sometimes just "shaping term". I'd suggest thinking of it as a reward function that happens to have the special property of being induced by a potential. Mentioning that to avoid any confusion when I talk about "potentials" etc. (and for the same reason might be better to name this variable something like "shaping")

File: src/epic/distances/divergence_free.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                ** 2,
+            )
+            l2_loss.backward()
+            optimizer.step()
+            optimizer.zero_grad()
+            losses.append(l2_loss.item())
+
+            # Early stopping if loss has stopped fluctuating
+            if len(losses) >= 500:
+                last_five_losses = losses[-500:]
```
Comment (ejnnr) [inline src/epic/distances/divergence_free.py:143]: These are the last 500 losses, not 5, right? (not sure which of the two you wanted)

File: src/epic/distances/divergence_free.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                if np.max(last_five_losses) - np.min(last_five_losses) < 1e-6:
+                    break
+            scheduler.step()
+        # plt.plot(losses)
+        # plt.show()
+
+        def canonical_reward_fn(state, action, next_state, done, /):
```
Comment (ejnnr) [inline src/epic/distances/divergence_free.py:150]: Suggestion for refactoring: with a few small changes (I think just the GPU/CPU handling), you could use this function above while optimizing the potential net (i.e. for the loss calculation). Would deduplicate a few lines of code and probably also make it clearer what's going on (you're returning exactly the function whose L2 norm you minimized)

File: src/epic/distances/pearson_mixin.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                n_samples_cov or self.default_samples_cov,
+            )
+
+        x_samples = x_canonical(state_cov_sample, action_cov_sample, next_state_cov_sample, done_cov_sample)
+        y_samples = y_canonical(state_cov_sample, action_cov_sample, next_state_cov_sample, done_cov_sample)
+
+        print(np.var(x_samples))
+        print(np.var(y_samples))
+
+        # handle cases with constant reward function
```
Comment (ejnnr) [inline src/epic/distances/pearson_mixin.py:55]: Just leaving a TODO comment that we probably want to handle this differently long term, but not exactly sure how yet and fine to leave as it is right now

File: src/epic/distances/pearson_mixin.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            )
+        else:
+            state_cov_sample, action_cov_sample, next_state_cov_sample, done_cov_sample = self.coverage_sampler.sample(
+                n_samples_cov or self.default_samples_cov,
+            )
+
+        x_samples = x_canonical(state_cov_sample, action_cov_sample, next_state_cov_sample, done_cov_sample)
+        y_samples = y_canonical(state_cov_sample, action_cov_sample, next_state_cov_sample, done_cov_sample)
+
+        print(np.var(x_samples))
```
Comment (ejnnr) [inline src/epic/distances/pearson_mixin.py:52]: Remove once done debugging and PR is ready

File: src/epic/modules.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# %%
```
Comment (ejnnr) [inline src/epic/modules.py:1]: Remove VS code notebook delimiters once PR is ready

File: tests/test_divergence_free_dist.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+    dist = divergence_free.DivergenceFree(
+        state_sampler=DummyGymStateSampler(space=state_space),
+        action_sampler=GymSpaceSampler(space=action_space),
+        discount_factor=1,
+    ).distance(x, y, n_samples_cov=500, n_samples_can=5000)
+
+    print(dist)
+
+    assert np.isclose(dist, 0, atol=2e-1)
```
Comment (ejnnr) [inline tests/test_divergence_free_dist.py:115]: I expected once things are working well, we should get significantly lower distance than 0.2 for linear potentials, lmk if that seems hard to achieve

File: tests/test_utils.py [hunk 10]
```diff
@@ -1,4 +1,5 @@
 import numpy as np
+import torch as torch
```
Comment (ejnnr) [inline tests/test_utils.py:2]: Any reason for this over just `import torch`?

Comment (anshradh) [reply tests/test_utils.py]: Ah haha, clearly just my brain on autopilot.

File: tests/test_divergence_free_dist.py (hunk 9 above)
Comment (anshradh) [reply tests/test_divergence_free_dist.py:125]: I think I've been getting somewhere between 0.05 and 0.2 (most runs below 0.1). I can maybe try increasing the depth of the network if it seems like it's definitely possible to do better.

File: src/epic/modules.py (hunk 8 above)
Comment (anshradh) [reply src/epic/modules.py]: Done.

File: src/epic/distances/pearson_mixin.py (hunk 7 above)
Comment (anshradh) [reply src/epic/distances/pearson_mixin.py]: Done.

File: src/epic/distances/pearson_mixin.py (hunk 6 above)
Comment (anshradh) [reply src/epic/distances/pearson_mixin.py:53]: Added a TODO.

File: src/epic/distances/divergence_free.py (hunk 5 above)
Comment (anshradh) [reply src/epic/distances/divergence_free.py]: Done, thanks for the suggestion.

File: src/epic/distances/divergence_free.py (hunk 4 above)
Comment (anshradh) [reply src/epic/distances/divergence_free.py]: Yeah this should probably just be renamed (and perhaps the early_stopping window should be configurable).

File: src/epic/distances/divergence_free.py (hunk 3 above)
Comment (anshradh) [reply src/epic/distances/divergence_free.py]: Changed to `shaping`

File: src/epic/distances/divergence_free.py (hunk 2 above)
Comment (anshradh) [reply src/epic/distances/divergence_free.py:77]: Yeah I think currently this isn't really adding much, figured it was good to have it just in case it improved things and since we'd probably want it at some point in the future.

File: src/epic/distances/divergence_free.py (hunk 1 above)
Comment (anshradh) [reply src/epic/distances/divergence_free.py]: Yeah makes sense, I think this isn't making a big difference right now (though I changed it to go over both the gelu nonlinearities as well now).

File: src/epic/distances/divergence_free.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        network) to minimize the L2 norm of the shaped reward function.
+        """
+        rew_fn = utils.multidim_rew_fn(reward_function)
+        n_samples_can = n_samples_can or self.default_samples_can
+        assert isinstance(n_samples_can, int)
+
+        net = nn.Sequential(
+            *[
+                nn.Flatten(),
+                nn.Linear(self.state_dim, max(self.state_dim * 4, 128)),
```
[the remark this answers is not in the document]
Comment (anshradh) [reply src/epic/distances/divergence_free.py]: Done.

File: src/epic/distances/divergence_free.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+        if coverage_sampler is None:
+            assert (
+                state_sampler is not None and action_sampler is not None
+            ), "If no coverage sampler is given, state and action samplers must be provided."
+        else:
+            assert (
+                state_sampler is None and action_sampler is None
+            ), "If a coverage sampler is given, state and action samplers will not be used."
+
+        coverage_sampler = coverage_sampler or samplers.ProductDistrCoverageSampler(action_sampler, state_sampler)
```
[the remark this answers is not in the document]
Comment (anshradh) [reply src/epic/distances/divergence_free.py]: Yeah I think this makes sense to remove.

File: src/epic/distances/divergence_free.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+import numpy.typing as npt
+import torch
+import torch.nn as nn
+import torch.optim as optim
+
+# import matplotlib.pyplot as plt
+
+from epic import samplers, types, utils
+from epic.distances import base, pearson_mixin
+from epic.modules import Residual
```
[the remark this answers is not in the document]
Comment (anshradh) [reply src/epic/distances/divergence_free.py]: Done.

File: tests/test_divergence_free_dist.py (hunk 9 above)
Comment (ejnnr) [reply tests/test_divergence_free_dist.py:125]: Well, in some sense it's definitely possible to do better: if our function class contains good approximations to any linear function, in particular to the potential, we should be able to get very close to zero distance. That said, I'm not sure how easy this is in practice (i.e. with imperfect optimization). If it turns out we can't get lower than this in such a simple setting though, that's bad news for the entire method. So I think worth figuring out what the limits are here before moving to more complicated cases.

One sanity check: if you use a linear model for the potential (i.e. just one `nn.Linear`), then I think you should be able to get very low distances (`< 1e-2` hopefully). If that's not working, we should look into that first, rather than scaling up the network (and in general, it seems somewhat weird if we'd need to scale up the network just to fit a linear function)

Comment (anshradh) [reply tests/test_divergence_free_dist.py:125]: Yeah that works (0.005 distance) (if I also increase `n_can_samples` to 25k). I'll try increasing the number of samples and go back to the MLP net to see if I can get it to work.

Comment (anshradh) [reply tests/test_divergence_free_dist.py:125]: I can get it to `< 5e-2` now, but going much lower seems like it's gonna be pretty difficult. I simplified to only use two linear layers with a gelu nonlinearity between them and also dramatically increased the number of canonicalization samples.

Comment (ejnnr) [reply tests/test_divergence_free_dist.py:125]: Ok, I was hoping this would work even better but sounds plausible that's hard. Probably makes sense to try more complicated potentials and see how things scale. If we can get to 5e-2 even in more complex settings, that's probably fine, whereas if things get much worse, we might have to come up with new ideas.

Two other things that might be good to test:
- How stable is optimization? I.e. if you train several times (with different optimization), do you get basically the same distance each time, or does it fluctuate a lot? This might be most visible in more complicated settings, and with reward functions that don't just differ by potential shaping.
- What if you canonicalize the zero reward function? In that case, the network can clearly fit the correct potential (by just setting all weights to zero). If this doesn't work any better than the `5e-2`, it indicates we might want to push a bit more after all (even more optimization)

File: src/epic/distances/divergence_free.py [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+                assert next_state_tensor.ndim == 1
+                state_tensor.unsqueeze_(-1)
+                next_state_tensor.unsqueeze_(-1)
+
+            net.to(device)
+
+            shaping = (self.discount_factor * net(next_state_tensor) - net(state_tensor)).squeeze(-1)
+
+            if not return_tensor:
+                return rew_fn(state, action, next_state, done) + utils.numpy_from_tensor(shaping)
```
Comment (ejnnr) [inline src/epic/distances/divergence_free.py:119]: Worth adding an `assert` that no broadcasting is happening here (i.e. both summands have the same `ndim`). E.g. if someone passes in a neural net as the reward function, and the neural net outputs a singleton dimension at the end, pytorch is going to add a singleton dimension to the front of the potential to match the number of dimensions, and then broadcast (which leads to the wrong result). Can be an incredibly annoying bug to find, so good to be defensive with shape asserts

File: tests/test_divergence_free_dist.py (hunk 9 above)
Comment (anshradh) [reply tests/test_divergence_free_dist.py:125]: - It seems reasonably stable, I'd say the fluctuations have been around the `1e-2` mark (Currently seems to be like `2.5e-2 +- 1e-2`. I'll run more extensive tests tomorrow to get a better idea of the variance in outcomes.
- Canonicalizing the zero reward function works, you can see other tests in the file that check for that in a few different cases (it just gets a distance of 0, since the samples are low variance enough that we treat both the zero reward function and the zero reward function with the potential as constant reward functions).

Comment (ejnnr) [reply tests/test_divergence_free_dist.py:125]: Ok, if I read the code correctly that means the zero reward functions gets canonicalized to variance less than `1e-2`, which seems pretty reasonable. If you do spend more time on improving the optimization, you could check just how far away the canonicalized reward is from exactly 0 (i.e. is the variance more like `5e-3` or `1e-5`). That seems like a good simple test case to tell whether you're making progress. But also very reasonable at this point to just scale things up and see what happens (and perhaps eventually move towards a pipeline with actual environments and reward models)

File: src/epic/distances/divergence_free.py (hunk 14 above)
Comment (anshradh) [reply src/epic/distances/divergence_free.py]: Done.

**Comment by anshradh** [comment]
Things TODO that probably should be done before this is used more extensively:

- Figure out how to get the training to be faster/more efificent. I think it's possible with the addition of LayerNorm (maybe this should be BatchNorm?) to the potential network, the learning rate can be increased and training can be done for less samples/steps. This is probably P0, since it can take > 1 hr to get results for reward functions that are neural networks themselves.
- Implementation should probably be more flexible to allow separation of the canonicalization and the distance metric. Should also maybe cache the canonicalized reward function after training so the network doesn't have to be trained again for comparing to a different reward function.
- More experiments comparing this with EPIC, both to verify for correctness and to notice places where they disagree
- Rename Repo :)
